### PR TITLE
Add support of the command ZDIFF and ZDIFFSTORE

### PR DESCRIPTION
**Pull request by MaheshMadushan**
Add support of the command ZDIFF and ZDIFFSTORE #1984
- [x] ZDIFF
- [x] ZDIFFSTORE

### CODE REVIEW

**Comment by torwig** [comment]
@MaheshMadushan Thank you for the contribution. 
You can format the codebase by running `./x.py format`.

**Comment by MaheshMadushan** [comment]
> @MaheshMadushan Thank you for the contribution. You can format the codebase by running `./x.py format`.

Hi @torwig , formatting issue fixed. Thank you.

**Comment by torwig** [comment]
@MaheshMadushan LGTM, nicely done!
Could you please add some edge-cases tests,? For example, if the first set is empty, doesn't exist, one of the other sets doesn't exist, etc.

**Comment by MaheshMadushan** [comment]
Hi @torwig . Thanks for the feedback. will add edge test cases for the feature.

File: src/types/redis_zset.cc [hunk 1]
```diff
@@ -930,4 +930,46 @@ rocksdb::Status ZSet::RandMember(const Slice &user_key, int64_t command_count,
   return rocksdb::Status::OK();
 }
 
+rocksdb::Status ZSet::Diff(const std::vector<Slice> &keys, MemberScores *members) {
+  std::vector<std::string> lock_keys;
```
Comment (git-hulk) [inline src/types/redis_zset.cc]: It should be fine to read the snapshot instead of locking keys.

**Comment by git-hulk** [comment]
Overall are good, one comment inline.

File: src/types/redis_zset.cc (hunk 1 above)
Comment (git-hulk) [reply src/types/redis_zset.cc]: Other places lock keys, so let's leave it as it is and fix them together.

**Review by git-hulk — DISMISSED** [review]

File: src/types/redis_zset.cc (hunk 1 above)
Comment (MaheshMadushan) [reply src/types/redis_zset.cc]: Hi @git-hulk, I really appreciate your feedback. I'd be happy to address this in the current PR if that works for you. Additionally, I was wondering if you could provide some more insights into the snapshotting process in kvrocks, specifically whether reading from snapshots is mutually exclusive? Your clarification would be incredibly helpful. Thanks a lot!

Comment (git-hulk) [reply src/types/redis_zset.cc]: @MaheshMadushan Thanks for your contribution. We don't need to use a lock while reading keys. For the read snapshot, please refer: https://github.com/apache/kvrocks/blob/unstable/src/types/redis_zset.cc#L839

>  specifically whether reading from snapshots is mutually exclusive? Your clarification would be incredibly helpful. 

It's not mutually exclusive to readers. You can regard it as a mechanism like MVCC, the snapshot won't be changed it's reading.

Comment (MaheshMadushan) [reply src/types/redis_zset.cc]: Thank you. Got it.

**Comment by torwig** [comment]
@git-hulk Are we waiting for additional test in this PR or it will be a separate one?

**Comment by git-hulk** [comment]
> @git-hulk Are we waiting for additional test in this PR or it will be a separate one?

@torwig ohh, sorry for missing your previous comment. @MaheshMadushan Could you please add the edge cases as @torwig mentioned in this PR?

**Comment by MaheshMadushan** [comment]
> > @git-hulk Are we waiting for additional test in this PR or it will be a separate one?
> 
> @torwig ohh, sorry for missing your previous comment. @MaheshMadushan Could you please add the edge cases as @torwig mentioned in this PR?

will add them asap.

File: src/types/redis_zset.cc [hunk 2]
```diff
@@ -931,14 +931,6 @@ rocksdb::Status ZSet::RandMember(const Slice &user_key, int64_t command_count,
 }
 
 rocksdb::Status ZSet::Diff(const std::vector<Slice> &keys, MemberScores *members) {
-  std::vector<std::string> lock_keys;
```
Comment (MaheshMadushan) [inline src/types/redis_zset.cc]: Hi @git-hulk , I have omitted the locking in the ZSet::Diff. is this what you had in mind?

Comment (git-hulk) [reply src/types/redis_zset.cc]: Yes, I'm good with this change. We can improve this in the future by using the same snapshot between keys in RangeByScore.

File: tests/gocase/unit/type/zset/zset_test.go [hunk 3]
```diff
@@ -1463,6 +1463,168 @@ func basicTests(t *testing.T, rdb *redis.Client, ctx context.Context, encoding s
...
+			{Score: 4, Member: "e"},
+		})
+		createZset(rdb, ctx, "zsetb", []redis.Z{
+			{Score: 1, Member: "b"},
+			{Score: 2, Member: "c"},
+			{Score: 4, Member: "f"},
+			{Score: 4, Member: "e"},
+		})
+		rdb.Del(ctx, "zsetc")
+		rdb.Del(ctx, "zsetc")
```
Comment (torwig) [inline tests/gocase/unit/type/zset/zset_test.go]: Why do you repeat the deletion?

**Comment by torwig** [comment]
@MaheshMadushan @git-hulk  I'm also thinking about one optimization: when the source zset is empty - I guess in this case we can make early return and there is no sense to traverse all the others zsets. Is it correct?

**Comment by git-hulk** [comment]
> @MaheshMadushan @git-hulk I'm also thinking about one optimization: when the source zset is empty - I guess in this case we can make early return and there is no sense to traverse all the others zsets. Is it correct?

Good point.

File: tests/gocase/unit/type/zset/zset_test.go (hunk 3 above)
Comment (MaheshMadushan) [reply tests/gocase/unit/type/zset/zset_test.go]: Thanks for pointing out. This is a mistake. will fix this.

**Comment by MaheshMadushan** [comment]
> @MaheshMadushan @git-hulk I'm also thinking about one optimization: when the source zset is empty - I guess in this case we can make early return and there is no sense to traverse all the others zsets. Is it correct?

Will fix this.

**Review by torwig — APPROVED** [review]

**Review by git-hulk — APPROVED** [review]

**Comment by git-hulk** [comment]
@MaheshMadushan Thanks for your contribution.

**Comment by MaheshMadushan** [comment]
@git-hulk @torwig Thank you for feedback.
